tests/benchmarks/runner.py

```diff
@@ -0,0 +1,242 @@
+
+
+BenchmarkIterations = NewType("BenchmarkIterations", int)
+
+
+@dataclass
+class BenchmarkSessionConfiguration:
+    iterations: BenchmarkIterations = BenchmarkIterations(1)
+
+
+@dataclass
+class BenchmarkSession:
+    """Benchmark session handling class."""
+
+    report: BenchmarkReport
+    runner: BenchmarkRunner
+    file_manager: BenchmarkFileManager
+    configurations: BenchmarkSessionConfiguration
+
+    def _update_configurations(self, **configs) -> None:
+        for config_name, config_value in configs.items():
+            setattr(self.configurations, config_name, config_value)
+
+    @contextmanager
+    def configure(self, **tmp_configurations):
+        """Temporarily replaces the session configurations.
+
+        See ``BenchmarkSessionConfiguration`` for available options.
+        """
+        original_configurations = {
+            config_name: getattr(self.configurations, config_name)
+            for config_name in tmp_configurations
+        }
+        self._update_configurations(**tmp_configurations)
+        yield None
+        self._update_configurations(**original_configurations)
+
+    def run(self, *runner_args, **parameters):
+        """Call ``self.runner`` with arguments and save the result.
+
+        Calls ``self.runner``: ``BenchmarkRunner``
+        with unpacked ``runner_args`` and ``parameters``
+        to generate a single benchmark report (``SingleRunReport``)
+        and append the result to the ``self.report``: ``BenchmarkReport``.
+
+        Use ``self.configure`` to use non-default configurations.
+        Configurable options should be handled by ``BenchmarkSessionConfiguration``
+        instead of having extra arguments in ``run`` methods.
+        So that all arguments of ``run`` can be directly passed to ``BenchmarkRunner``.
+        """
+        for i_iter in range(1, self.configurations.iterations + 1):
+            single_report = self.runner(*runner_args, **parameters)
+            self.report.append(single_report)
+            if i_iter == self.configurations.iterations:
+                return single_report.output
+
+    def save(self):
+        """Save the accumulated benchmark results in the container (``self.report``).
+
+        Calls  ``self.file_manager.save`` with ``self.report``,
+        so that ``BenchmarkFileManager`` can dump the result into a file.
+        """
+        self.file_manager.save(self.report)
+
+
+def create_benchmark_runner_factory(
+    runner_type: type[BenchmarkRunner] = SimpleRunner,
+) -> Factory:
+    session_providers = ProviderGroup()
+    session_providers[BenchmarkSession] = BenchmarkSession
+    session_providers[BenchmarkReport] = BenchmarkReport
+    session_providers[BenchmarkRunner] = runner_type
+    session_providers[BenchmarkFileManager] = SimpleFileManager
+    session_providers[BenchmarkSessionConfiguration] = BenchmarkSessionConfiguration
+    return Factory(env_providers, session_providers)
```

Comment:
Can you explain what this is doing?

Reply:
It is just a helper to collect all providers to build ``BenchmarkSession`` with dependency injection module.
I also added a doc string.